sample/weighted.go

```diff
@@ -0,0 +1,98 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this code is governed by a BSD-style
+// license that can be found in the LICENSE file
+
+package sample
+
+import "math/rand"
+
+// Weighted provides sampling without replacement from a collection of items with
+// weighted probabilites.
+type Weighted struct {
+	weights []float64
+	heap    []float64
+	src     *rand.Rand
+}
+
+// NewWeighted returns a Weighted for the weights w. Elements of w are modified
+// during sampling, so a copy should be passed if the user needs to keep the values.
+// If src is nil, rand.Rand is used as the random source.
+func NewWeighted(w []float64, src *rand.Rand) Weighted {
+	s := Weighted{
+		weights: w,
+		heap:    make([]float64, len(w)),
+	}
+	s.reset()
+	return s
+}
+
+func (s Weighted) reset() {
+	copy(s.heap, s.weights)
+	for i := len(s.heap) - 1; i > 0; i-- {
+		// Sometimes 1-based counting makes sense.
+		s.heap[((i+1)>>1)-1] += s.heap[i]
+	}
+}
+
+// Len returns the number of items held by the Weighted, including items
+// already taken.
+func (s Weighted) Len() int { return len(s.weights) }
+
+// Take returns an index from the Weighted based on the Weighted weighting, the
```

Comment:
Take returns an index from the Weighted with probability proportional to the weight of the item. The weight of the item is then set to zero. Take returns false if there are no items remaining.